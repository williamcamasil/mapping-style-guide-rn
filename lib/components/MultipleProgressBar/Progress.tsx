import React, { useMemo } from 'react';
import { StyleProp, ViewStyle } from 'react-native';

import { LinearGradient } from 'expo-linear-gradient';

import { useStyles } from '../../hooks';
import { withTheme } from '../../theme';
import { AppThemeType } from '../../theme/default';
import { ColorPaletesType } from '../../tokens';

type ProgressPropsType = {
  progress: number;
  theme: AppThemeType;
  startColor: ColorPaletesType;
  endColor: ColorPaletesType;
  style?: StyleProp<ViewStyle>;
};

const startLinearGradient = { x: 0, y: 0 };
const endLinearGradient = { x: 1, y: 0 };

const Progress: React.FC<ProgressPropsType> = ({
  progress, theme, startColor, endColor, style,
}) => {
  const colorGradient = useMemo(
    () => [theme.colors[startColor], theme.colors[endColor]],
    [endColor, startColor, theme],
  );

  const styleBar = useStyles(
    () => [
      {
        flex: progress,
        height: theme.spacings.sNano,
        borderRadius: theme.spacings.sNano / 2,
      },
      style,
    ],
    [progress, style, theme.spacings.sNano],
  );

  return (
    <LinearGradient
      testID="progress-linear-gradient"
      colors={colorGradient}
      start={startLinearGradient}
      end={endLinearGradient}
      style={styleBar}
    />
  );
};

Progress.defaultProps = {
  progress: 0,
};

export default withTheme(Progress);
