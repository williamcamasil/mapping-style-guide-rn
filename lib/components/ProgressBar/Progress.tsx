import React, { useMemo } from 'react';

import { LinearGradient } from 'expo-linear-gradient';

import { useStyles } from '../../hooks';
import { withTheme } from '../../theme';
import { AppThemeType } from '../../theme/default';
import { ColorPaletesType } from '../../tokens';

type ProgressPropsType = {
  progress: number;
  theme: AppThemeType;
  startColor?: ColorPaletesType;
  endColor?: ColorPaletesType;
};

const startLinearGradient = { x: 0, y: 0 };
const endLinearGradient = { x: 1, y: 0 };

const Progress: React.FC<ProgressPropsType> = ({
  progress, theme, startColor, endColor,
}) => {
  const colorGradient = useMemo(
    () => [theme.colors[startColor ?? 'primary600'], theme.colors[endColor ?? 'primaryMain']],
    [endColor, startColor, theme],
  );

  const styleBar = useStyles(
    () => [
      {
        width: progress >= 0 && progress <= 100 ? `${progress}%` : 0,
        height: '100%',
        borderRadius: 50,
      },
    ],
    [progress],
  );

  return (
    <LinearGradient
      testID="progress-linear-gradient"
      colors={colorGradient}
      start={startLinearGradient}
      end={endLinearGradient}
      style={styleBar}
    />
  );
};

Progress.defaultProps = {
  progress: 0,
};

export default withTheme(Progress);
