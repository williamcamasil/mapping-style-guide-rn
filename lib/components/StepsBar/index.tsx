import React from 'react';
import { StyleSheet, View, Text as RNText } from 'react-native';

import { useTextStyles } from '../../hooks';
import { withTheme } from '../../theme';
import { AppThemeType } from '../../theme/default';
import ProgressBar, { ProgressBarType } from '../ProgressBar';
import Text from '../Text';

type StepsBarPropsType = {
  step: number;
  totalSteps: number;
  theme: AppThemeType;
  size?: ProgressBarType['size'];
};

const style = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  progress: {
    flex: 1,
  },
});

const StepsBar: React.FC<StepsBarPropsType> = ({
  step, totalSteps, theme, size,
}) => {
  const current = Math.max(step, 0);
  const total = Math.max(current, totalSteps);
  const progress = total > 0 ? (current / total) * 100 : 0;

  const stepContainerStyled = useTextStyles(
    () => [
      {
        marginLeft: theme.spacings.sSmall,
      },
    ],
    [theme],
  );

  return (
    <View style={style.container}>
      <View style={style.progress}>
        <ProgressBar progress={progress} size={size} />
      </View>
      <View style={stepContainerStyled}>
        <Text variant="caption" color="neutralGray600" lineHeight="small">
          Passo
          {' '}
          <RNText testID="steps-bar-step-text" style={theme.typography.weights.bold}>
            {step}
          </RNText>
          {' '}
          de
          {' '}
          <RNText testID="steps-bar-total-text" style={theme.typography.weights.bold}>
            {totalSteps}
          </RNText>
          {' '}
        </Text>
      </View>
    </View>
  );
};

StepsBar.defaultProps = {
  size: 'medium',
};

export default withTheme(StepsBar);
