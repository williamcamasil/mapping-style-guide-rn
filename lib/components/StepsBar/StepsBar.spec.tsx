import React from 'react';

import { render, within } from '@testing-library/react-native';

import StepsBar from '.';
import Progress from '../ProgressBar/Progress';

describe('StepsBar events', () => {
  it('Should verify render the component', () => {
    const result = render(<StepsBar step={1} totalSteps={5} size="medium" />);

    expect(result).toBeDefined();
  });

  it('Should verify if progressbar will render with correct value', () => {
    const step = 1;
    const totalSteps = 5;

    const result = render(<StepsBar step={step} totalSteps={totalSteps} size="medium" />);

    const progressBarContainer = result.getByTestId('progress-bar-container');
    const progressBar = within(progressBarContainer).UNSAFE_getByType(Progress);

    const onePercent = (100 / totalSteps) * step;

    expect(progressBar.props).toHaveProperty('progress', onePercent);
  });

  it('Should verify if value of step will to render', () => {
    const result = render(<StepsBar step={1} totalSteps={5} size="medium" />);

    const stepText = result.getByTestId('steps-bar-step-text');

    expect(stepText.props.children).toBe(1);
  });

  it('Should verify if value of totalSteps will to render', () => {
    const result = render(<StepsBar step={1} totalSteps={5} size="medium" />);

    const stepText = result.getByTestId('steps-bar-total-text');

    expect(stepText.props.children).toBe(5);
  });

  it('Should not throw error if totalSteps is zero', () => {
    const step = 0;
    const totalSteps = 0;

    const result = render(<StepsBar step={step} totalSteps={totalSteps} size="medium" />);

    const stepText = result.getByTestId('steps-bar-total-text');

    expect(stepText.props.children).toBe(0);
  });
});

describe('Snapshot to StepsBar', () => {
  it('Snapshot initial', () => {
    const result = render(<StepsBar step={1} totalSteps={5} size="medium" />).toJSON();

    expect(result).toMatchSnapshot();
  });

  it('Snapshot fifty percent of progress', () => {
    const result = render(<StepsBar step={2.5} totalSteps={5} size="medium" />).toJSON();

    expect(result).toMatchSnapshot();
  });

  it('Snapshot hundred percent of progress', () => {
    const result = render(<StepsBar step={5} totalSteps={5} size="medium" />).toJSON();

    expect(result).toMatchSnapshot();
  });
});
