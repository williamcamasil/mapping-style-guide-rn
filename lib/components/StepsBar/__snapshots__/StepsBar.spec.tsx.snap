// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot to StepsBar Snapshot fifty percent of progress 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
    }
  }
>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "#F7F7F8",
          "borderColor": "#F7F7F8",
          "borderRadius": 6,
          "borderWidth": 2,
          "height": 12,
          "justifyContent": "center",
          "overflow": "hidden",
        }
      }
      testID="progress-bar-container"
    >
      <View
        colors={
          [
            "#F29B72",
            "#E95814",
          ]
        }
        end={
          {
            "x": 1,
            "y": 0,
          }
        }
        start={
          {
            "x": 0,
            "y": 0,
          }
        }
        style={
          {
            "borderRadius": 50,
            "height": "100%",
            "width": "50%",
          }
        }
        testID="progress-linear-gradient"
      />
    </View>
  </View>
  <View
    style={
      {
        "marginLeft": 20,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        {
          "color": "#676D79",
          "flexShrink": 1,
          "fontFamily": "Averta CY",
          "fontSize": 13,
          "fontWeight": "normal",
          "letterSpacing": -0.02,
          "lineHeight": 13,
        }
      }
    >
      Passo
       
      <Text
        style={
          {
            "fontFamily": "Averta CY",
            "fontWeight": "bold",
          }
        }
        testID="steps-bar-step-text"
      >
        2.5
      </Text>
       
      de
       
      <Text
        style={
          {
            "fontFamily": "Averta CY",
            "fontWeight": "bold",
          }
        }
        testID="steps-bar-total-text"
      >
        5
      </Text>
       
    </Text>
  </View>
</View>
`;

exports[`Snapshot to StepsBar Snapshot hundred percent of progress 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
    }
  }
>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "#F7F7F8",
          "borderColor": "#F7F7F8",
          "borderRadius": 6,
          "borderWidth": 2,
          "height": 12,
          "justifyContent": "center",
          "overflow": "hidden",
        }
      }
      testID="progress-bar-container"
    >
      <View
        colors={
          [
            "#F29B72",
            "#E95814",
          ]
        }
        end={
          {
            "x": 1,
            "y": 0,
          }
        }
        start={
          {
            "x": 0,
            "y": 0,
          }
        }
        style={
          {
            "borderRadius": 50,
            "height": "100%",
            "width": "100%",
          }
        }
        testID="progress-linear-gradient"
      />
    </View>
  </View>
  <View
    style={
      {
        "marginLeft": 20,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        {
          "color": "#676D79",
          "flexShrink": 1,
          "fontFamily": "Averta CY",
          "fontSize": 13,
          "fontWeight": "normal",
          "letterSpacing": -0.02,
          "lineHeight": 13,
        }
      }
    >
      Passo
       
      <Text
        style={
          {
            "fontFamily": "Averta CY",
            "fontWeight": "bold",
          }
        }
        testID="steps-bar-step-text"
      >
        5
      </Text>
       
      de
       
      <Text
        style={
          {
            "fontFamily": "Averta CY",
            "fontWeight": "bold",
          }
        }
        testID="steps-bar-total-text"
      >
        5
      </Text>
       
    </Text>
  </View>
</View>
`;

exports[`Snapshot to StepsBar Snapshot initial 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
    }
  }
>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "#F7F7F8",
          "borderColor": "#F7F7F8",
          "borderRadius": 6,
          "borderWidth": 2,
          "height": 12,
          "justifyContent": "center",
          "overflow": "hidden",
        }
      }
      testID="progress-bar-container"
    >
      <View
        colors={
          [
            "#F29B72",
            "#E95814",
          ]
        }
        end={
          {
            "x": 1,
            "y": 0,
          }
        }
        start={
          {
            "x": 0,
            "y": 0,
          }
        }
        style={
          {
            "borderRadius": 50,
            "height": "100%",
            "width": "20%",
          }
        }
        testID="progress-linear-gradient"
      />
    </View>
  </View>
  <View
    style={
      {
        "marginLeft": 20,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        {
          "color": "#676D79",
          "flexShrink": 1,
          "fontFamily": "Averta CY",
          "fontSize": 13,
          "fontWeight": "normal",
          "letterSpacing": -0.02,
          "lineHeight": 13,
        }
      }
    >
      Passo
       
      <Text
        style={
          {
            "fontFamily": "Averta CY",
            "fontWeight": "bold",
          }
        }
        testID="steps-bar-step-text"
      >
        1
      </Text>
       
      de
       
      <Text
        style={
          {
            "fontFamily": "Averta CY",
            "fontWeight": "bold",
          }
        }
        testID="steps-bar-total-text"
      >
        5
      </Text>
       
    </Text>
  </View>
</View>
`;
